File: frontend/src/services/memory.ts
import api from './api';

interface ApiResponse<T> {
  code: number;
  message: string;
  data: T;
}

export interface MemoryRecord {
  id: string;
  question: string;
  answer: string;
  created_at?: string;
  source_note?: string;
}

export const memoryApi = {
  /** 保存一条问答记忆 */
  save: async (question: string, answer: string, source_note?: string): Promise<boolean> => {
    const response = await api.post<ApiResponse<{ saved: boolean }>>('/memory/save', {
      question,
      answer,
      source_note,
    });
    return ((response as unknown as ApiResponse<{ saved: boolean }>).data)?.saved ?? false;
  },

  /** 列出已存记忆（按创建时间倒序） */
  list: async (limit = 200, offset = 0): Promise<MemoryRecord[]> => {
    const response = await api.get<ApiResponse<MemoryRecord[]>>('/memory/list', {
      params: { limit, offset },
    });
    return (response as unknown as ApiResponse<MemoryRecord[]>).data ?? [];
  },

  /** 删除一条记忆 */
  remove: async (id: string): Promise<boolean> => {
    const response = await api.delete<ApiResponse<{ deleted: boolean }>>(
      `/memory/${encodeURIComponent(id)}`
    );
    return ((response as unknown as ApiResponse<{ deleted: boolean }>).data)?.deleted ?? false;
  },

  /** 记忆条数 */
  count: async (): Promise<number> => {
    const response = await api.get<ApiResponse<{ count: number }>>('/memory/count');
    return ((response as unknown as ApiResponse<{ count: number }>).data)?.count ?? 0;
  },
};

File: backend/app/routers/memory.py
"""问答记忆 API（持久化记忆 - MVP）"""
from typing import Optional
from fastapi import APIRouter
from pydantic import BaseModel, Field
from app.services.memory_service import memory_service
from app.utils.logger import logger

router = APIRouter(prefix="/api/memory", tags=["memory"])


class MemorySaveRequest(BaseModel):
    question: str = Field(..., description="用户问题")
    answer: str = Field(..., description="已验证的回答")
    source_note: Optional[str] = Field(None, description="来源备注（可选）")


class MemorySearchRequest(BaseModel):
    query: str = Field(..., description="查询问题")
    top_k: int = Field(default=3, ge=1, le=20)
    threshold: Optional[float] = Field(default=None, ge=0.0, le=1.0)


@router.post("/save")
async def save_memory(req: MemorySaveRequest):
    ok = memory_service.save_qa(req.question, req.answer, req.source_note)
    logger.info(f"保存问答记忆: {req.question[:50]} -> {ok}")
    return {
        "code": 200 if ok else 500,
        "message": "保存成功" if ok else "保存失败",
        "data": {"saved": ok},
    }


@router.get("/search")
async def search_memory(query: str, top_k: int = 3, threshold: Optional[float] = None):
    hits = memory_service.search(query, top_k=top_k, threshold=threshold)
    return {
        "code": 200,
        "message": "success",
        "data": hits,
        "total": len(hits),
    }


@router.post("/search")
async def search_memory_post(req: MemorySearchRequest):
    hits = memory_service.search(req.query, top_k=req.top_k, threshold=req.threshold)
    return {
        "code": 200,
        "message": "success",
        "data": hits,
        "total": len(hits),
    }


@router.get("/count")
async def count_memory():
    return {
        "code": 200,
        "message": "success",
        "data": {"count": memory_service.count()},
    }


@router.get("/list")
async def list_memory(limit: int = 100, offset: int = 0):
    records = memory_service.list_all(limit=limit, offset=offset)
    return {
        "code": 200,
        "message": "success",
        "data": records,
        "total": len(records),
    }


@router.delete("/{mem_id}")
async def delete_memory(mem_id: str):
    ok = memory_service.delete(mem_id)
    return {
        "code": 200 if ok else 404,
        "message": "删除成功" if ok else "记忆不存在或删除失败",
        "data": {"deleted": ok},
    }
